examples/generative/ddim.py

```diff
@@ -0,0 +1,839 @@
+    crop_size = tf.minimum(height, width)
+    image = tf.image.crop_to_bounding_box(
+        data["image"],
+        (height - crop_size) // 2,
+        (width - crop_size) // 2,
+        crop_size,
+        crop_size,
+    )
+
+    # resize and clip
+    # for image downsampling it is important to turn on antialiasing
+    image = tf.image.resize(image, size=[image_size, image_size], antialias=True)
+    return tf.clip_by_value(image / 255.0, 0.0, 1.0)
+
+
+def prepare_dataset(split):
+    # the validation dataset is shuffled as well, because data order matters
+    # for the KID estimation
+    return (
+        tfds.load(dataset_name, split=split, shuffle_files=True)
+        .map(preprocess_image, num_parallel_calls=tf.data.AUTOTUNE)
+        .cache()
+        .repeat(dataset_repetitions)
+        .shuffle(10 * batch_size)
+        .batch(batch_size, drop_remainder=True)
+        .prefetch(buffer_size=tf.data.AUTOTUNE)
+    )
+
+
+# load dataset
+train_dataset = prepare_dataset("train[:80%]+validation[:80%]+test[:80%]")
+val_dataset = prepare_dataset("train[80%:]+validation[80%:]+test[80%:]")
+
+"""
+## Kernel inception distance
+
+[Kernel Inception Distance (KID)](https://arxiv.org/abs/1801.01401) is an image quality
+metric which was proposed as a replacement for the popular
+[Frechet Inception Distance (FID)](https://arxiv.org/abs/1706.08500).
+I prefer KID to FID because it is simpler to
+implement, can be estimated per-batch, and is computationally lighter. More details
+[here](https://keras.io/examples/generative/gan_ada/#kernel-inception-distance).
+
+In this example, the images are evaluated at the minimal possible resolution of the
+Inception network (75x75 instead of 299x299), and the metric is only measured on the
+validation set for computational efficiency. We also limit the number of sampling steps
+at evaluation to 5 for the same reason.
+
+Since the dataset is relatively small, we go over the train and validation splits
+multiple times per epoch, because the KID estimation is noisy and compute-intensive, so
+we want to evaluate only after many iterations, but for many iterations.
+
+"""
+
+
+class KID(keras.metrics.Metric):
+    def __init__(self, name="kid", **kwargs):
+        super().__init__(name=name, **kwargs)
+
+        # KID is estimated per batch and is averaged across batches
+        self.kid_tracker = keras.metrics.Mean()
+
+        # a pretrained InceptionV3 is used without its classification layer
+        # transform the pixel values to the 0-255 range, then use the same
+        # preprocessing as during pretraining
+        self.encoder = keras.Sequential(
+            [
+                layers.InputLayer(input_shape=(image_size, image_size, 3)),
+                layers.Rescaling(255.0),
+                layers.Resizing(height=kid_image_size, width=kid_image_size),
+                layers.Lambda(keras.applications.inception_v3.preprocess_input),
+                keras.applications.InceptionV3(
```

Comment:
What motivates the choice of InceptionV3 (which is quite heavy)?

Reply:
A smaller network, like a pretrained MobileNet, could probably work as well. But InceptionV3 is what is commonly used in FID and KID metrics, they even have it in their names. I reduce its resolution from 299 to 75 to be lighter. I did the same in a GAN code example: https://github.com/keras-team/keras-io/blob/master/examples/generative/gan_ada.py#L221